### ISSUE TITLE
[Firefox] Scripts randomly stop working

### ISSUE THREAD

**Issue by darkred** [Tampermonkey/tampermonkey#477]
Tried TM 4.5.5633 stable and 4.5.5648beta in FF 57.0.1 x64 in win10.

The issue I have is this:

- _(initially TM works ok)_
- I close and restart FF:  in the restored pages (session restore), sometimes the scripts run in the page, but frequently other they don't. (although the counter in the toolbar button still displays each matching counter) . Let's suppose that they do work.
- I disable TM via its toolbar button menu(the 1st entry), reload the page(F5) to make sure the scripts are not working anymore, and then re-enable it via the toolbar button:
- no scripts are working anymore (although the counter in the toolbar button still displays each matching counter) . 
- At that time, no matter how many times 
I either disable+re-enable TM the toolbar button, 
or I disable+re-enable TM via about:addons,
they don't run in the page anymore.
- The only solution I've found is, when this happens,
to close+re-open FF (at that time, the scripts will still not be working), and disable+re-enable TM via the toolbar button. <-- _but even this, it doesn't fix the problem every time, I may have to disable+re-enable TM multiple times and/or restart FF in order to fix this ._
And, If I try to disable it then, the issue re-occurs (almost) every single time.

I haven't managed to recreate this with a fresh FF profile.

Here is a log from Browser Console (TM 4.5.5633 stable)
as I disable TM via the toolbar button (having only TM enabled):

```
Try to load layout default  layout.js:3
No matching message handler for the given recipient. MessageChannel.jsm:703
Storage: use chromeStorage  storage.js:19
Unchecked lastError value: Error: Could not establish connection. Receiving end does not exist. ExtensionCommon.jsm:407
Storage: test -> start  storage.js:11
Storage: test -> set after 502ms  storage.js:11
Storage: test -> get after 504ms  storage.js:11
Storage: test -> remove after 1010ms  storage.js:12
Unchecked lastError value: Error: Promised response from onMessage listener at addListener@moz-extension://eacdd3d4-ef20-4256-ae58-3eb85f85fa83/rea/common.js:3:291 went out of scope ExtensionCommon.jsm:407
```

and, as I re-enable TM via the toolbar button:

```
Try to load layout default  layout.js:3
Unchecked lastError value: Error: Promised response from onMessage listener at addListener@moz-extension://eacdd3d4-ef20-4256-ae58-3eb85f85fa83/rea/common.js:3:291 went out of scope  ExtensionCommon.jsm:407
Ignoring response to aborted listener for 575-4
Ignoring response to aborted listener for 576-4
Ignoring response to aborted listener for 577-4
Ignoring response to aborted listener for 578-4
Ignoring response to aborted listener for 579-4
Unchecked lastError value: Error: Could not establish connection. Receiving end does not exist.  ExtensionCommon.jsm:407
Storage: use chromeStorage  storage.js:19
Storage: test -> start  storage.js:11
Storage: test -> set after 502ms  storage.js:11
Storage: test -> get after 503ms  storage.js:11
Storage: test -> remove after 1005ms  storage.js:12


```

I don't have any special TM settings enabled
(especially, in Settings | General, the "Auto reload pages" setting remains disabled, as by default)

If there's anything more I can provide to help, please ask.

------

I have ~200 userscripts installed.
For reference, with the same number of scripts in Greasemonkey 4.0 or 4.1beta6 
I didn't have such an issue.

---

My other addons are uBlock Origin 1.14.20, Reddit Enhancement Suite 5.9.3beta and Stylus, but even if I disable them, the issue still occurs.

**Comment by derjanb** [comment]
I can not reproduce this issue. For me scripts are working fine after Tampermonkey is enabled again.
Can you please create a new Firefox profile and test with only Tampermonkey installed and one script? Does that work? If so does it still work with all scripts installed (you can export and import them at the "Utilities"tab).

**Comment by darkred** [comment]
Yes. of course. So:

---

I tried the following test with a clean profile FF 57.0.2 x64 profile and TM 4.5.5633:

- installed this one script [Mouseover Popup Image Viewer](https://greasyfork.org/en/scripts/404-mouseover-popup-image-viewer)
- open https://www.google.gr/search?q=firefox&tbm=isch
- mouseover a thumbnail for a large image: _the full image loads as a popup, as expected._
- disable TM via the toolbar button, refresh page (F5), re-enable TM via the toolbar button
- repeated the above 20 times: the issue never happens.

---

Now, I tried exporting (ZIP) my scripts and importing them to the fresh profile:
- Open once more https://www.google.gr/search?q=firefox&tbm=isch
- repeated the previous procedure 20 times: the issue never happens.

---

But, I tried opening another URL (I'd prefer to send this to you privately) in that fresh profile,
and upon e.g. 5 times **it will surely occur** - it happened to me 2 times in 20 tries or so.

---

&nbsp;
So, would you mind if I sent you my ZIP archive together with that URL, via pm in the TM forum,
to take a look at it?

**Comment by derjanb** [comment]
Sure. You can also send an email to support<ät>tampermonkey.net

**Comment by darkred** [comment]
I just sent you an email. Thank you.

**Comment by darkred** [comment]
With the new version 4.5.5660 the issue is fixed! Thanks a lot!

**Comment by darkred** [comment]
I spoke too soon: 
although with with the new version I may no longer recreate the issue I describe in the OP,
nevertheless, unfortunately,  five times since yesterday all scripts have stopped working, 
and it occurs **randomly**, just by itself, i.e. _without_ me disabling TM it via the toolbar button menu(=TM remains enabled in the toolbar button menu).

The above happens to me in my everyday profile - I couldn't recreate it in the fresh profile with only TM and my ZIP, like in the issue I described here. This might be because, with the new version, the scripts stop working, not in a consistent, reproducible manner, but it seems to happen randomly.
&nbsp;

For reference, with the new version, the scripts stopped working in the URL I sent you, as well as : 

- in GitHub pages, e.g. https://github.com 
_(most apparent of all my installed scripts that stop working, is the script [GitHub Custom Navigation](https://greasyfork.org/scripts/20830) which adds navigational buttons on top the page),_
- e.g. http://www.metal-archives.com/bands/Kamelot/166. 
_(most apparent is that the [script](https://greasyfork.org/scripts/5751) that splits the Reviews column, stops working)_ and 
- https://greasyfork.org/ 
_(most apparent is the script that changes the site layout: [[TS] Citrus GFork](https://greasyfork.org/en/scripts/4336-ts-citrus-gfork) stops working)_

**Comment by arogl** [comment]
I have a similar issue, I have tested both the latest TM and beta in Firefox and a clean profile.

If I add 3 scripts for IMDB, they work when I install TM.

When I close FF and then restart and go back to IMDB, TM says 3 scripts running, but no changes are injected into IMDB web page.

I am trying to look at debugging and will attach when I find anything.

**Comment by derjanb** [comment]
> Unfortunately, I spoke too soon:
although with with the new version I may no longer recreate the issue I describe in the OP,
nevertheless, five times since yesterday, all scripts have stopped working,
and it occurs randomly, just by itself, i.e. without me disabling TM it via the toolbar button menu

Hmm, this sounds like it could be related to #485

**Comment by darkred** [comment]
> I have problems to reporduce this problem. Are the scripts working again after a refresh?

No, reloading the page (F5) doesn't help - restarting Firefox helped, though.
&nbsp;
   
What has happened three times so far since yesterday is that the scripts may be working in a tab in the main/normal window (even after a reload/F5 on the page) 
but not in another tab, always in the same main/normal window!
 
_Using FF 58 and latest TM beta 4.6.5694) with my everyday FF profile._
What I describe above has happened me in tabs of the same normal window with e.g.
https://rarbgproxy.org/torrents.php
http://www.imdb.com/title/tt0076759/
https://www.rottentomatoes.com/m/blade_runner_2049/

**while at the same time** scripts would run ok in other tabs!! :
i.e. all the scripts running in https://github.com/ would still work ok,
and [MPIV](https://greasyfork.org/en/scripts/404-mouseover-popup-image-viewer) would still work in tabs different than the one that had imdb.com !!

Unfortunately there is no relevant entry in the Browser Console when this happens.

I hope this helps.

Restarting Firefox fixed the issue.

**Comment by darkred** [comment]
Using FF 58 and latest TM beta 4.6.5694) 

I'd like to add that restarting Firefox fixes the problem indeed, but not always on the first restart: 
e..g. today my script [Instagram - visible images counter](https://greasyfork.org/en/scripts/11635-instagram-visible-images-counter) _(which runs on instagram profile pages, such as https://www.instagram.com/instagram/  )_ stopped working, and I restarted Firefox 5 times (with Firefox's session restore always having that instagram URL as the selected tab) 
and then the script worked again.

**Comment by darkred** [comment]
The issue still occurs (scripts randomly stop working) in latest TM beta 4.6.5703 _using FF58_,

**Comment by darkred** [comment]
Also, a thing I notice very frequently:
when a script randomly stops in a page,
e.g. [[TS] Citrus GFork](https://greasyfork.org/en/scripts/4336-ts-citrus-gfork) in https://greasyfork.org/en
or [Mouseover Popup Image Viewer])https://greasyfork.org/en/scripts/404-mouseover-popup-image-viewer) in random pages,
then, if, at session restore, the focused tab is the same as that when the script had stopped working and on which I had closed FF from the last session, 
then the issue is not fixed no matter how many times I restart Firefox.
For the issue to be fixed, I have to focus a different tab and then close and restart FF.

**Comment by darkred** [comment]
I get the following same error in Browser Console whenever scripts stop working in a tab -although scripts might still work in other tabs- :
_(at that time the counter in the toolbar button has no value for that tab)_ 
 
 ```
Error: WebExtension context not found!                                         ExtensionParent.jsm:855:13
```

Using FF 59 and TM 4.6.5709

**Comment by darkred** [comment]
In order to rule out that the cause the issue was caused because the TM userscript database in Firefox had might become corrupted for some reason,
I even uninstalled Tampermonkey (i.e. incl. all installed scripts) in FF, 
and restored them from a ZIP backup of TM for Chrome (the same scripts, ~200),
but unfortunately, the issue still occurs: scripts randomly stop working in some tabs.

I still get the same error in Browser Console when this happens: 

```
Error: WebExtension context not found!               ExtensionParent.jsm:861:13
```

Using TM 4.6 in FF 60 and Chrome 66.

**Comment by TempACount** [comment]
I can confirm this happens on Chrome as well- been having this issue for months (but not for the maybe two years prior to then, with the same scripts). My assumption is that at some point an update to TamperMonkey introduced an element that, under certain conditions, causes the program to improperly configure or process.

**Comment by darkred** [comment]
I've noticed the following TM 4.7.5788 beta in FF 60, which was occuring very commonly
In e.g. this page https://greasyfork.org/en/users/2160-darkred?sort=updated
commonly all scripts would stop working - most apparently the [[TS] Citrus GFork](https://greasyfork.org/en/scripts/4336-ts-citrus-gfork) which completely reformats the page.

From the ZIP I've sent you, @derjanb, here are the active scripts for the above URL:
![2018-06-07_220146](https://user-images.githubusercontent.com/723651/41121005-e87f99a0-6a9f-11e8-9f7a-b232398817c9.jpg)

When all scripts would randomly stop working , no matter how many times I would reload the page, the problem wouldn't get fixed.

Of all the 14 scripts (in the screenshot) running, only 1 was running at `document-start`, [Greasebox](https://arantius.com/misc/greasemonkey/greasebox.user.js) .

I found that if I disabled it, then when I reloaded the page, all scripts would work ok, again.

I tried it multiple times (in other URLs too, where Greasebox was the only `document-start` script), and once the issue occurred, 
it was consistently fixed whenever I disabled that script,
and occured again when I re-enabled it.

I'm positive that this issue is a duplicate of #211.
Maybe, @derjanb you would like to take a look at the 'Greasebox' script, too.

---

FWIW I was getting these two errors in Firefox's Browser Console,
in both cases, i.e. when the issue occured, and when it was fixed:

```
Error: WebExtension context not found!  ExtensionParent.jsm:861:13
Unchecked lastError value: Error: Could not establish connection. Receiving end does not exist.  background.js:3}
```

PS. The issue doesn't occur in Chrome for me, never.

**Comment by darkred** [comment]
In addition to the above post, I've noticed that, even with GreaseBox is disabled,
if at session restore the focused tab is the above page, https://greasyfork.org/en/users/2160-darkred?sort=updated
then in most cases, the scripts in that tab (mainly '[TS] Citrus GFork') don't work,
and the only way to fix is to close Firefox while having focus on a different URL, and restart Firefox.

i.e. similar to my previous comment: https://github.com/Tampermonkey/tampermonkey/issues/477#issuecomment-370857498.

In other words, I believe that **the issue is additionally related to #485**, as you have already suggested.

**Comment by arogl** [comment]
Found another issue, on restart in FF 64/beta/nightly Tampermonkey BETA or normal does not always start on browser startup. I eventually narrowed it down to this script

[W.A.R. Links Checker Premium](http://usa.x10host.com/mybb/showthread.php?tid=23)

Turn the script off and Tampemonkey starts every time. 

On chrome with the same scripts and settings this does not cause an issue. How can I help debug the startup issue?

**Comment by phuclv90** [comment]
It's quite unfortunate that this 4-year-old bug is still not yet solved. But looks like it only appears on Windows. I've never seen the same issue in Firefox for mac

**Comment by wvffle** [comment]
> But looks like it only appears on Windows.

I experience it on linux.

**Comment by derjanb** [comment]
Closing in favor of #1016
